File: chap06/python_files/Sort/Quick.py
from Sort.Sort_template import Sort
from Stack.int_stack import IntStack

class Quick(Sort):
  def sort(self):
    def check(obj, return_value):
      print(f"{obj.__name__} is Failed.") if return_value == -1 else print("", end="")

    start = 0
    size = self.get_list_len()
    lstack = IntStack(size - start + 1)
    rstack = IntStack(size - start + 1)
    l_v = lstack.Push(start)
    check(lstack, l_v) 
    r_v = rstack.Push(size-1)
    check(rstack, r_v)
    x = super().list
    while(lstack._IsEmpty()==False):
      pl = start = lstack.Pop()
      check(lstack, pl)
      pr = end = rstack.Pop()
      check(rstack, pr)
      middle_value = x[int((pl + pr)//2)] 
      while(pl <= pr):
        while(x[pl] < middle_value): 
          pl += 1
        while(x[pr] > middle_value):
          pr -= 1
        if(pl <= pr):
          x[pl], x[pr] = x[pr], x[pl]
          pl += 1
          pr -= 1
      if(start < pr):
        l_v = lstack.Push(start)
        check(lstack, l_v)
        r_v = rstack.Push(pr)
        check(rstack, r_v)
      if(pl < end):
        l_v = lstack.Push(pl)
        check(lstack, l_v)
        r_v = rstack.Push(end)
        check(rstack, r_v)
    del lstack
    del rstack


  def get_list_len(self):
    return super().size   
 

  def recursive_sort(self, start, end):
    def partition(a_list, start, end):
      x = a_list
      left = start; right = end;
      middle_value = x[int((left + right) / 2)]
      while(left <= right):
        while(middle_value > x[left]):
          left += 1
        while(middle_value < x[right]):
          right -= 1
        if(left <= right):
          x[left], x[right] = x[right], x[left]
          left += 1
          right -= 1
      return left, right

    x = self.list
    left, right = partition(x, start, end)
    if(start < right): 
      self.recursive_sort(start, right)
    if(end > left):
      self.recursive_sort(left, end)

File: chap06/source_files/quick.c
#include"quick.h"
#include"IntStack.h"

void print_values(int a[], int s, int e)
{
  int i;

  for(i = s; i < e; i++)
    printf("%d ", a[i]);
  putchar('\n');
}

void quick(int a[], int left, int right)
{
  int pl;
  int pr;

  partition(a, &pl, &pr, left, right);
  if(left < pr) quick(a, left, pr);
  if(right > pl) quick(a, pl, right);
}

void quick_iterative(int a[], int left, int right)
{
  IntStack lstack;
  IntStack rstack;
  int pl, pr, x;

  Initialize(&lstack, right - left + 1);
  Initialize(&rstack, right - left + 1);

  Push(&lstack, left);
  Push(&rstack, right);

  while(!IsEmpty(&lstack)){
    pl = (Pop(&lstack, &left), left);
    pr = (Pop(&rstack, &right), right);
    x = a[(left + right) / 2];
    do{
      while(a[pl] < x) pl++;
      while(a[pr] > x) pr--;
      if(pl <= pr){
        swap(int, a[pl], a[pr]);
        pl++;
        pr--;
      }
    }while(pl <= pr);

   if(left < pr){
     Push(&lstack, left);
     Push(&rstack, pr);
   } 
   if(pl < right){
     Push(&lstack, pl);
     Push(&rstack, right);
   }
  }

  Terminate(&lstack);
  Terminate(&rstack);
}

int main(void)
{
  int i, nx;
  int *x;
  int *x2;

  puts("퀵 정렬");
  printf("요소 개수 : ");
  scanf("%d", &nx);
  x = malloc(nx * sizeof(int));
  x2 = calloc(nx, sizeof(int));
  for(i = 0; i < nx; i++){
    printf("x[%d]: ", i);
    scanf("%d", &x[i]);
    x2[i] = x[i];
  }
  quick(x, 0, nx-1);
  puts("오름차순으로 정렬했습니다.");
  print_values(x, 0, nx);
  
  quick_iterative(x2, 0, nx - 1);
  puts("오름차순으로 정렬했습니다.");
  print_values(x2, 0, nx);
  free(x);
  free(x2);
  return 0;
}
